drivers/eeprom/eeprom_mchp_xec.c

```diff
@@ -0,0 +1,326 @@
+#define XEC_EEPROM_MODE_ACTIVATE		BIT(0)
+
+/* EEPROM Status Register */
+#define XEC_EEPROM_STS_TRANSFER_COMPL		BIT(0)
+
+/* EEPROM Execute Register - Transfer size bit position */
+#define XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS	(24)
+
+/* EEPROM Execute Register - Commands */
+#define XEC_EEPROM_EXC_CMD_READ			0x00000U
+#define XEC_EEPROM_EXC_CMD_WRITE		0x10000U
+#define XEC_EEPROM_EXC_CMD_READ_STS		0x20000U
+#define XEC_EEPROM_EXC_CMD_WRITE_STS		0x30000U
+
+/* EEPROM Execute Register - Address mask */
+#define XEC_EEPROM_EXC_ADDR_MASK		0x7FFU
+
+/* EEPROM Status Byte */
+#define XEC_EEPROM_STS_BYTE_WIP			BIT(0)
+#define XEC_EEPROM_STS_BYTE_WENB		BIT(1)
+
+/* EEPROM Read/Write Transfer Size */
+#define XEC_EEPROM_PAGE_SIZE			32U
+#define XEC_EEPROM_TRANSFER_SIZE_READ		XEC_EEPROM_PAGE_SIZE
+#define XEC_EEPROM_TRANSFER_SIZE_WRITE		XEC_EEPROM_PAGE_SIZE
+
+#define XEC_EEPROM_DELAY_US			500U
+#define XEC_EEPROM_DELAY_BUSY_POLL_US		50U
+#define XEC_EEPROM_XFER_COMPL_RETRY_COUNT	10U
+
+struct eeprom_xec_regs {
+	uint32_t mode;
+	uint32_t execute;
+	uint32_t status;
+	uint32_t intr_enable;
+	uint32_t password;
+	uint32_t unlock;
+	uint32_t lock;
+	uint32_t _reserved;
+	uint8_t buffer[XEC_EEPROM_PAGE_SIZE];
+};
+
+struct eeprom_xec_config {
+	struct eeprom_xec_regs * const regs;
+	size_t size;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pcfg;
+#endif
+};
+
+K_MUTEX_DEFINE(lock);
+
+static void eeprom_xec_execute_reg_set(struct eeprom_xec_regs * const regs,
+						uint32_t transfer_size, uint32_t command,
+						uint16_t eeprom_addr)
+{
+	uint32_t temp = command + (eeprom_addr & XEC_EEPROM_EXC_ADDR_MASK);
+
+	if (transfer_size != XEC_EEPROM_PAGE_SIZE) {
+		temp += (transfer_size << XEC_EEPROM_EXC_TRANSFER_SZ_BITPOS);
+	}
+	regs->execute = temp;
+}
+
+static void eeprom_xec_data_buffer_read(struct eeprom_xec_regs * const regs,
+					uint8_t transfer_size, uint8_t *destination_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			*destination_ptr = regs->buffer[count];
+			destination_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_data_buffer_write(struct eeprom_xec_regs * const regs,
+							uint8_t transfer_size, uint8_t *source_ptr)
+{
+	uint8_t count;
+
+	if (transfer_size <= XEC_EEPROM_PAGE_SIZE) {
+		for (count = 0; count < transfer_size; count++) {
+			regs->buffer[count] = *source_ptr;
+			source_ptr++;
+		}
+	}
+}
+
+static void eeprom_xec_wait_transfer_compl(struct eeprom_xec_regs * const regs)
+{
+	uint8_t sts = 0;
+	uint8_t retry_count = 0;
+
+	k_sleep(K_USEC(XEC_EEPROM_DELAY_US));
+
+	do {
+		if (retry_count >= XEC_EEPROM_XFER_COMPL_RETRY_COUNT) {
+			LOG_ERR("XEC EEPROM retry count exceeded");
+			break;
+		}
+		k_sleep(K_USEC(XEC_EEPROM_DELAY_BUSY_POLL_US));
+
+		sts = XEC_EEPROM_STS_TRANSFER_COMPL & regs->status;
+		retry_count++;
+
+	} while (!sts);
+
+	/* Clear the appropriate status bits */
+	regs->status = XEC_EEPROM_STS_TRANSFER_COMPL;
```

Comment:
You may want to update `regs->status` conditionally: `if (sts)`. 
Cause, you may break out of the above while loop after max retries.